docs/rules/no-implicit-service-injection-argument.md

```diff
@@ -0,0 +1,40 @@
+# no-implicit-service-injection-argument
+
+:wrench: The `--fix` option on the [command line](https://eslint.org/docs/user-guide/command-line-interface#fixing-problems) can automatically fix some of the problems reported by this rule.
+
+Disallow omitting the service name argument when injecting a service.
+
+Some developers may prefer to be more explicit about what service is being injected instead of relying on the implicit and potentially costly lookup/normalization of the service from the property name.
```

Comment:
@rwjblue is this the right justification for this rule?

Reply:
@runspired thoughts?

Reply:
It's generally the right motivation, though in truth my feelings here have shifted with time.

With classic syntax, the implicit felt extra magical. With native class syntax and decorators, the propertyName being the key is a lot more clear (and understandable) I feel.

I think at this point what I want is for an error if the propertyName would be different from resolved name, I'm not sure if we can give the rule enough context for that or not. The key point is to enforce that it would work with the strict-resolver package if desired.

@scalvert or @stefanpenner may have good feedback here.

Reply:
@runspired thanks for the context. I agree with your goal of ensuring the service injection property matches the resolved name to avoid the need for normalization. I'm thinking of proceeding with this plan:

1. Merge `no-implicit-service-injection-argument` as-is, always requiring the service injection argument. This would be intended as an optional (non-recommended) rule for people who prefer this style.
   * In the future, we could potentially make it smarter/configurable about when to require the service injection argument.
2. Create a new rule `require-normalized-service-injections` (or similar name) which would both validate that the service being injected actually exists, and that the name it's injected with matches its resolved name (so that it works with ember-strict-resolver). This can be implemented once it's practical for the lint rule to understand the full service resolution logic (including how to find services in addons).